File: WC3/C++_homework/Car/Car.cpp
#include "Point.h"
#include "Car.h"

Car::Car(double capacity, double consumption, const Point& location, const std::string& model) {
    this->fuelAmount = 0;
    this->fuelCapacity = capacity;
    this->fuelConsumption = consumption;
    this->location = location;
    this->model = model;
}
Car::~Car() {}

double Car::getFuelAmount() const {
    return this->fuelAmount;
}
double Car::getFuelCapacity() const {
    return this->fuelCapacity;
}
double Car::getFuelConsumption() const {
    return this->fuelConsumption;
}
const Point& Car::getLocation() const {
    return this->location;
}
const std::string& Car::getModel() const {
    return this->model;
}

void Car::drive(const Point& destination) {
    double fuelNeeded = this->location.Point::distance(destination) * this->getFuelConsumption();

    if ( this->fuelAmount < fuelNeeded ) {
        throw OutOfFuel();
    } else {
        this->location = destination;
        this->fuelAmount -= fuelNeeded;
        std::cout << "Car has arrived to " << destination << std::endl;
    }
}
void Car::drive(double x, double y) {
    Point destination = Point(x,y);
    
    this->drive(destination);
}

void Car::refill(double fuel) {
    if ( fuel < 0 ) {
        fuel = 0;
    }
    if ( this->fuelAmount + fuel > this->fuelCapacity ) {
        this->fuelAmount = this->fuelCapacity;
        throw ToMuchFuel();
    } else {
        this->fuelAmount += fuel;
    }
}

std::ostream& operator<<(std::ostream& out, const Car& car) {
    out << "Car model is " << car.getModel() << std::endl << "Current fuel = " << car.getFuelAmount() << "/" << car.getFuelCapacity() << std::endl << "Fuel consumption = " << car.getFuelConsumption() << std::endl << "Car location = " << car.getLocation();
    return out;
}

File: WC3/Java_homework/Car/Car.java
public class Car implements Cloneable {
    private double fuelAmount;
    private double fuelCapacity;
    private double fuelConsumption;
    Point location;
    String model;
    
    public Car(double capacity, double consumption, Point location, String model) {
        this.fuelAmount = 0;
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.location = location;
        this.model = model;
    }
    public Car() {
        this.fuelAmount = 0;
        this.fuelCapacity = 1;
        this.fuelConsumption = 0.1;
        this.location = new Point(0.0,0.0);
        this.model = "Unknown model";
    }
    
    public double getFuelAmount() {
        return this.fuelAmount;
    }
    
    public double getFuelCapacity() {
        return this.fuelCapacity;
    }
    public double getFuelConsumption() {
        return this.fuelConsumption;
    }
    public Point getLocation() {
        return this.location;
    }
    public String getModel() {
        return this.model;
    }
    
    public void drive(Point destination) throws OutOfFuelException {
        double fuelNeeded = this.location.distance(destination) * this.fuelConsumption;
        
        if ( this.fuelAmount < fuelNeeded ) {
            throw new OutOfFuelException();
        }
        this.location = destination;
        this.fuelAmount -= fuelNeeded;
        System.out.println("Car has arrived to " + destination);
        
    }
    public void drive(double x, double y) throws OutOfFuelException {
        Point destination = new Point(x,y);
        this.drive(destination);
    }
    
    public void refill(double fuel) throws ToMuchFuelException {
        if ( fuel > this.fuelCapacity ) {
            throw new ToMuchFuelException();
        }
        this.fuelAmount += fuel;
    }
    public void refill() {
        this.fuelAmount = this.fuelCapacity;
    }
    
    public String toString() {
        return "\n" + "Car model: " + this.model + "\n" + "Fuel capacity: " + this.fuelAmount + "/" + this.fuelCapacity + "\n" + "Fuel consumption: " + this.fuelConsumption + "\n" + "location: " + this.location + "\n";
    }
}